### PR TITLE
Fix vertical zoom actions disable track height lock, Add action "SWS/NF: Toggle track height - and vertical zoom actions obey track height lock"

### LINKED ISSUE #1221: Vertical zoom actions disable track height lock
reminder for:
https://forum.cockos.com/showthread.php?t=205372

I think issue is in `Zoom.cpp/VertZoomRange()` (maybe other places as well, didn't check)
https://github.com/reaper-oss/sws/blob/221e764c213efb8f0f0def93e07e021c17d153ab/Zoom.cpp#L131
https://github.com/reaper-oss/sws/blob/221e764c213efb8f0f0def93e07e021c17d153ab/Zoom.cpp#L263

as this disables the height lock.
https://forum.cockos.com/showthread.php?p=2202082#post2202082

### PR DESCRIPTION
**Pull request by nofishonfriday**
fixes #1221, implements #966 (optionally)
The toggle action was added so [users who prefer the current behaviour](https://forum.cockos.com/showthread.php?p=2091347#post2091347) (Xenakios track height actions and SWS vertical zoom actions ignore track height locking) can have that too.

+revert adding actions "Xenakios/SWS/NF: Set selected tracks heights to A/B, respect height locked tracks" (added in v2.12.0pre)
+refactor "Xenakios/SWS: Set selected tracks heights to A/B" (use COMMAND_T::user)

### CODE REVIEW

File: Zoom.cpp [hunk 1]
```diff
@@ -243,8 +263,23 @@ void VertZoomRange(int iFirst, int iNum, bool* bZoomed, bool bMinimizeOthers, bo
...
+				if (!SWS_IsTrackHeightLocked(tr)) 
+					SetMediaTrackInfo_Value(tr, "I_HEIGHTOVERRIDE", g_i0);
+			}
+		}
+		else
+		{
+			for (int i = 0; i <= GetNumTracks(); i++) 
+			{
+				MediaTrack* tr = CSurf_TrackFromID(i, false);
+				SetMediaTrackInfo_Value(tr, "I_HEIGHTOVERRIDE", minTrackHeight);
```
Comment (nofishonfriday) [inline Zoom.cpp:280]: This is wrong, but I don't know how to fix currently.

**Comment by nofishonfriday** [comment]
@cfillion 
Would you perhaps help me with this (or tackle yourself if you prefer)?
I'm stuck and don't know how to do it properly. :(

**Comment by cfillion** [comment]
I'm a bit out of the loop, but wouldn't restoring `B_HEIGHTLOCK` to its original value after setting `I_HEIGHTOVERRIDE` do the trick?

**Comment by nofishonfriday** [comment]
The problem (for me) is that the current code resets the `I_HEIGHTOVERRIDE`
https://github.com/reaper-oss/sws/blob/23871fdc873aae987f5a5db4596c35009fc5c77a/Zoom.cpp#L247
`GetSetMediaTrackInfo(CSurf_TrackFromID(i, false), "I_HEIGHTOVERRIDE", &g_i0);`

which disables the track locking, resp. locking can't be set anymore for these tracks.
(From the [API doc](https://www.extremraym.com/cloud/reascript-doc/#SetMediaTrackInfo_Value):)
> B_HEIGHTLOCK : bool * : track height lock (must set I_HEIGHTOVERRIDE before locking)

**Comment by cfillion** [comment]
How about explicitly setting the height to `trackHeight` if the track is locked + obey track height lock is enabled?

(The snippet below doesn't have the "if" part)

```cpp
diff --git a/Zoom.cpp b/Zoom.cpp
index 2f27cbef..ced779d3 100644
--- a/Zoom.cpp
+++ b/Zoom.cpp
@@ -216,6 +216,7 @@ void VertZoomRange(int iFirst, int iNum, bool* bZoomed, bool bMinimizeOthers, bo
 
                int iZoom = VZOOM_RANGE+1;
                int iHeight;
+               int trackHeight = 0;
                do
                {
                        iZoom--;
@@ -225,7 +226,7 @@ void VertZoomRange(int iFirst, int iNum, bool* bZoomed, bool bMinimizeOthers, bo
                                MediaTrack* tr = CSurf_TrackFromID(i+iFirst, false);
                                if (TcpVis(tr))
                                {
-                                       int trackHeight = GetTrackHeightFromVZoomIndex(tr, iZoom);
+                                       trackHeight = GetTrackHeightFromVZoomIndex(tr, iZoom);
 
                                        int envHeight = 0;
                                        if (i < (int)envelopeHeights.size())
@@ -244,7 +245,7 @@ void VertZoomRange(int iFirst, int iNum, bool* bZoomed, bool bMinimizeOthers, bo
 
                // Reset custom track sizes
                for (int i = 0; i <= GetNumTracks(); i++)
-                       GetSetMediaTrackInfo(CSurf_TrackFromID(i, false), "I_HEIGHTOVERRIDE", &g_i0);
+                       GetSetMediaTrackInfo(CSurf_TrackFromID(i, false), "I_HEIGHTOVERRIDE", &trackHeight);
                *ConfigVar<int>("vzoom2") = iZoom;
                TrackList_AdjustWindows(false);
                UpdateTimeline();
```

**Pull request closed without merging**